File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/c5127f5691b4754aa4e976d93cb6a983.java
import java.util.*;
import java.io.*;

public class c5127f5691b4754aa4e976d93cb6a983_nc {
    static final int md = 998244353;
    static final long llinf = 1e18 + 100;
    static final int maxn = (int) 1e5 + 100;
    static final int inf = (int) 1e9 + 100;

    static String s;
    static List<Integer>[] o = new ArrayList[2];
    static List<Integer> ind = new ArrayList<>();
    static int[][] nxt;
    static int it;
    static int[] id = new int[maxn];
    static int[] mas = new int[maxn];
    static int[][] seg = new int[maxn][2];
    static Pair[] len = new Pair[maxn];

    static {
        o[0] = new ArrayList<>();
        o[1] = new ArrayList<>();
        nxt = new int[maxn][2];
        for (int i = 0; i < maxn; i++) {
            len[i] = new Pair(0, 0);
        }
    }

    static class Pair {
        int fr, sc;

        Pair(int fr, int sc) {
            this.fr = fr;
            this.sc = sc;
        }
    }

    static boolean can(int n) {
        if (n == 1) return true;

        int IT = it;
        int x = n / 2, y = (n + 1) / 2;
        if (!can(x) || !can(y)) return false;

        int p0 = nxt[it][0], p1 = nxt[it][1];
        if (p0 == s.length() || ind.get(p0) + x - 1 >= o[0].size())
            p0 = s.length();
        else
            p0 = o[0].get(ind.get(p0) + x - 1);
        if (p1 == s.length() || ind.get(p1) + y - 1 >= o[1].size())
            p1 = s.length();
        else
            p1 = o[1].get(ind.get(p1) + y - 1);
        p0 = Math.min(p0, p1);
        if (p0 == s.length()) return false;

        it = p0 + 1;
        return true;
    }

    static void gen(int l, int r) {
        if (l == r) return;
        int m = (l + r - 1) / 2;
        gen(l, m);
        gen(m + 1, r);
        int a = l, b = m + 1, c = l;
        while (a <= m && b <= r) {
            if (s.charAt(it) == '0') {
                mas[c++] = id[a++];
            } else {
                mas[c++] = id[b++];
            }
            it++;
        }
        while (a <= m) {
            mas[c++] = id[a++];
        }
        while (b <= r) {
            mas[c++] = id[b++];
        }
        for (int i = l; i <= r; i++) {
            id[i] = mas[i];
        }
    }

    static boolean go(int n, boolean cert) {
        it = 0;
        if (!cert) return can(n);
        it = 0;
        can(n);
        assert it == s.length();

        it = 0;
        for (int i = 0; i < n; i++) id[i] = i;
        gen(0, n - 1);
        assert it == s.length();
        int[] p = new int[n];
        for (int i = 0; i < n; i++) p[id[i]] = i;
        System.out.println(n);
        for (int i : p) System.out.print((i + 1) + " ");
        System.out.println();
        System.exit(0);
        return true;
    }

    static void solve() {
        for (int i = 2; i < maxn; i++) {
            int x = i / 2, y = i - x;
            len[i].fr = len[x].fr + len[y].fr + x;
            len[i].sc = len[x].sc + len[y].sc + x + y - 1;
            seg[i][0] = seg[x][0] + seg[y][0] + x;
            seg[i][1] = seg[x][1] + seg[y][1] + y;
        }
        Scanner scanner = new Scanner(System.in);
        s = scanner.next();
        int[] counter = new int[2];
        nxt = new int[s.length() + 1][2];
        for (int i = 0; i < s.length(); i++) {
            int c = s.charAt(i) - '0';
            counter[c]++;
            ind.add(o[c].size());
            o[c].add(i);
        }
        nxt[s.length()][0] = nxt[s.length()][1] = s.length();
        for (int i = s.length() - 1; i >= 0; i--) {
            for (int t = 0; t < 2; t++)
                nxt[i][t] = nxt[i + 1][t];
            nxt[i][s.charAt(i) - '0'] = i;
        }
        int l = 0, r = (int) 1e5 + 10;
        while (r - l > 1) {
            int m = (l + r) / 2;
            if (go(m, false)) l = m;
            else r = m;
        }
        go(l, true);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int ts = 1;
        for (int its = 1; its <= ts; its++) {
            solve();
        }
    }
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/c5127f5691b4754aa4e976d93cb6a983.py
import sys
import random
from collections import defaultdict
from itertools import accumulate

class Modular:
    def __init__(self, value=0, mod=998244353):
        self.value = value % mod
        self.mod = mod

    def __add__(self, other):
        return Modular(self.value + other.value, self.mod)

    def __sub__(self, other):
        return Modular(self.value - other.value, self.mod)

    def __mul__(self, other):
        return Modular(self.value * other.value, self.mod)

    def __truediv__(self, other):
        return self * Modular(pow(other.value, self.mod - 2, self.mod), self.mod)

    def __eq__(self, other):
        return self.value == other.value

    def __lt__(self, other):
        return self.value < other.value

    def __int__(self):
        return self.value

    def __repr__(self):
        return str(self.value)

def can(n):
    global it
    if n == 1:
        return True

    x = n // 2
    y = (n + 1) // 2
    if not can(x) or not can(y):
        return False

    p0 = nxt[it][0]
    p1 = nxt[it][1]
    if p0 == len(s) or ind[p0] + x - 1 >= len(o[0]):
        p0 = len(s)
    else:
        p0 = o[0][ind[p0] + x - 1]
    
    if p1 == len(s) or ind[p1] + y - 1 >= len(o[1]):
        p1 = len(s)
    else:
        p1 = o[1][ind[p1] + y - 1]
    
    p0 = min(p0, p1)
    if p0 == len(s):
        return False
    
    it = p0 + 1
    return True

def gen(l, r):
    global it
    if l == r:
        return
    
    m = (l + r - 1) // 2
    gen(l, m)
    gen(m + 1, r)
    a, b, c = l, m + 1, l
    while a <= m and b <= r:
        if s[it] == '0':
            mas[c] = id[a]
            a += 1
        else:
            mas[c] = id[b]
            b += 1
        c += 1
        it += 1
    
    while a <= m:
        mas[c] = id[a]
        a += 1
        c += 1
    while b <= r:
        mas[c] = id[b]
        b += 1
        c += 1
    
    for i in range(l, r + 1):
        id[i] = mas[i]

def go(n, cert):
    global it
    it = 0
    if not cert:
        return can(n)
    
    it = 0
    can(n)
    assert it == len(s)

    it = 0
    id[:] = range(n)
    gen(0, n - 1)
    assert it == len(s)
    p = [0] * n
    for i in range(n):
        p[id[i]] = i
    print(n)
    print(' '.join(str(i + 1) for i in p))
    sys.exit(0)

maxn = 100100
s = ""
o = [[], []]
ind = []
nxt = []
it = 0
id = [0] * maxn
mas = [0] * maxn

def solve():
    global s, it, o, ind, nxt
    for i in range(2, maxn):
        x = i // 2
        y = i - x
        len[i] = (len[x][0] + len[y][0] + x, len[x][1] + len[y][1] + x + y - 1)
        for t in range(2):
            seg[i][t] = seg[x][t] + seg[y][t]
        seg[i][0] += x
        seg[i][1] += y

    s = input().strip()
    counter = [0, 0]
    nxt = [[0, 0] for _ in range(len(s) + 1)]
    ind = [0] * len(s)
    for i in range(len(s)):
        c = int(s[i])
        counter[c] += 1
        ind[i] = len(o[c])
        o[c].append(i)
    
    nxt[len(s)][0] = nxt[len(s)][1] = len(s)
    for i in range(len(s) - 1, -1, -1):
        for t in range(2):
            nxt[i][t] = nxt[i + 1][t]
        nxt[i][int(s[i])] = i
    
    l, r = 0, 100010
    while r - l > 1:
        m = (l + r) // 2
        if go(m, False):
            l = m
        else:
            r = m
    go(l, True)

if __name__ == "__main__":
    input = sys.stdin.read
    data = input().splitlines()
    sys.stdin = open("../a.in", "r")
    sys.stdout = open("../a.out", "w")
    
    solve()
